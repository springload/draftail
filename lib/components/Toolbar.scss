.#{$DRAFTAIL}Toolbar {
  position: relative;
  padding: $controls-spacing;
  background-color: $draftail-editor-chrome;
  color: $draftail-editor-chrome-text;
  border-radius: $draftail-toolbar-radius;
  border-bottom: $draftail-toolbar-border-bottom;

  .#{$DRAFTAIL}Editor--focus & {
    // Browsers that do not support this will simply not have this added UI.
    // stylelint-disable-next-line plugin/no-unsupported-browser-features
    position: sticky;
    top: 0;
    z-index: $draftail-toolbar-z-index;
  }
}
