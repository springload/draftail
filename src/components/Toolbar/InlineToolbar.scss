@use "../../api/constants" as *;

$button-size: 1.5rem;

.#{$DRAFTAIL}ToolbarButton--pin {
  position: absolute;
  top: 0;
  inset-inline-end: 0;
  padding: 0;
  min-width: 1.5rem;
  height: 1.5rem;
  color: $draftail-editor-chrome;
  background-color: $draftail-editor-chrome-text;
  // Remove once we drop support for Safari 14.
  // stylelint-disable-next-line property-disallowed-list
  border-top-left-radius: 0;
  border-start-start-radius: 0;
  // Remove once we drop support for Safari 14.
  // stylelint-disable-next-line property-disallowed-list
  border-bottom-right-radius: 0;
  border-end-end-radius: 0;
  border: $draftail-toolbar-border-bottom;

  .#{$DRAFTAIL}Icon {
    width: 0.75rem;
    height: 0.75rem;
  }

  .#{$DRAFTAIL}ToolbarButton__label {
    vertical-align: top;
  }

  &::before {
    margin-bottom: 2px;
  }

  &::after {
    margin-bottom: 8px;
  }
}

.#{$DRAFTAIL}Toolbar--pin {
  padding-inline-end: $button-size;
}
