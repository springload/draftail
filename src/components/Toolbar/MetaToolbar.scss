@use "../../api/constants" as *;

.#{$DRAFTAIL}MetaToolbar {
  position: relative;
  display: flex;
  justify-content: flex-end;
  padding: $controls-spacing;
  background-color: $draftail-editor-background;
  color: $draftail-editor-chrome;
  border-top: $draftail-contrast-outline;
}
