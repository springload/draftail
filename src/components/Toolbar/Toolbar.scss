@use "../../api/constants" as *;

.#{$DRAFTAIL}Toolbar {
  position: relative;
  padding: $controls-spacing;
  background-color: $draftail-editor-chrome;
  color: $draftail-editor-chrome-text;
  border-radius: $draftail-toolbar-radius;
  border-bottom: $draftail-toolbar-border-bottom;

  .#{$DRAFTAIL}Editor--focus & {
    position: sticky;
    top: 0;
    z-index: $draftail-toolbar-z-index;
  }
}
