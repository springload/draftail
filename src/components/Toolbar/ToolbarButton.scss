@use "sass:color";
@use "../../api/constants" as *;

$button-active-color: color.adjust(
  $color: $draftail-editor-chrome-active,
  $alpha: -0.9,
);
$button-active-border-color: color.adjust(
  $color: $draftail-editor-chrome-active,
  $alpha: -0.8,
);
$button-font-size: 1rem;

.#{$DRAFTAIL}ToolbarButton {
  // Guarantee buttons are a predictable round-number size.
  $size: $button-font-size + $button-spacing * 2 + $draftail-base-spacing;
  @include font-smoothing();

  display: inline-block;
  padding: $button-spacing;
  min-width: $size;
  max-height: $size;
  border-radius: $draftail-toolbar-radius;
  background: transparent;
  color: inherit;
  border: 1px solid transparent;
  outline: $draftail-contrast-outline;
  font-size: $button-font-size;
  font-weight: 600;
  cursor: pointer;
  user-select: none;

  &--active {
    background-color: $button-active-color;
    border: 1px solid $button-active-border-color;

    @media (forced-colors: active) {
      // Force the text to CanvasText as forced colors automatically
      // adds a Canvas outline behind text.
      background: Highlight;
      color: HighlightText;
      border-color: currentColor;
    }
  }

  &__label {
    display: inline-block;
    vertical-align: middle;
    height: 1em;

    @media (forced-colors: active) {
      color: CanvasText;
    }
  }
}
