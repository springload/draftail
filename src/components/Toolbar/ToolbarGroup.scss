@use "../../api/constants" as *;

.#{$DRAFTAIL}ToolbarGroup {
  display: inline-block;

  &::before {
    content: "";
    display: inline-block;
    width: 0.0625rem;
    height: 1rem;
    vertical-align: middle;
    margin: 0 $controls-spacing * 0.5;
    background-color: $draftail-editor-chrome-accent;
  }

  &:first-of-type::before {
    display: none;
  }

  .#{$DRAFTAIL}Editor--readonly & {
    opacity: $draftail-editor-readonly-opacity;
  }
}
