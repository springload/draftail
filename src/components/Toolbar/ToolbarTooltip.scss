@use "sass:math";
@use "sass:color";

// Code initially taken from Balloon.css
// See https://github.com/kazzkiq/balloon.css.
//
// Variables
// -----------------------------------------

$balloon-bg: color.adjust(
  $color: #111,
  $alpha: -0.1,
) !default;
$balloon-text-color: $color-white;
$arrow-height: 6px;
$arrow-width: 18px;
$arrow-offset: 5px;
$balloon-radius: $draftail-toolbar-tooltip-radius;
$balloon-transition-duration: $draftail-toolbar-tooltip-duration;
$balloon-transition-delay: $draftail-toolbar-tooltip-delay;
$balloon-font-size: 0.875em;

//
// Mixins
// -----------------------------------------

@mixin arrow($color, $direction) {
  border: math.div($arrow-width, 2) solid transparent;
  @if $direction == up {
    border-top-width: 0;
    border-bottom: $arrow-height solid $color;
  } @else if $direction == down {
    border-bottom-width: 0;
    border-top: $arrow-height solid $color;
  } @else {
    @error "Unknown direction #{$direction}.";
  }
}

@mixin balloon-position($direction) {
  $prop: null;
  $factor: null;

  @if $direction == up {
    $prop: top;
    $factor: -1;
  } @else if $direction == down {
    $prop: bottom;
    $factor: 1;
  } @else {
    @error "Unknown direction #{$direction}.";
  }

  &::after,
  &::before {
    inset-inline-start: 50%;
    #{$prop}: 100%;
    transform: translate(
      calc(-50% * var(--draftail-text-direction)),
      $factor * 10px
    );
  }

  &:hover {
    &::after,
    &::before {
      transform: translate(calc(-50% * var(--draftail-text-direction)), 0);
    }
  }

  &::after {
    margin-#{$prop}: $arrow-offset + $arrow-height;
  }

  &::before {
    @include arrow($color: $balloon-bg, $direction: $direction);
    margin-#{$prop}: $arrow-offset;
  }
}

//
// Styles
// -----------------------------------------

[data-draftail-balloon] {
  position: relative;

  // Fixing iOS Safari event issue.
  // More info at: https://goo.gl/w8JF4W
  cursor: pointer;

  &::before,
  &::after {
    position: absolute;
    z-index: $draftail-toolbar-tooltip-z-index;
    opacity: 0;
    pointer-events: none;
  }

  &::before {
    content: "";
  }

  &::after {
    @include font-smoothing();

    background: $balloon-bg;
    border-radius: $balloon-radius;
    outline: $draftail-contrast-outline-modal;
    color: $balloon-text-color;
    content: attr(aria-label);
    padding: 0.5em 1em;
    white-space: pre;
    font-size: $balloon-font-size;
  }

  &:hover {
    &::before,
    &::after {
      opacity: 1;
      transition: all $balloon-transition-duration ease-out
        $balloon-transition-delay;
    }
  }
}

[data-draftail-balloon="up"] {
  @include balloon-position($direction: up);
}

[data-draftail-balloon="down"] {
  @include balloon-position($direction: down);
}
