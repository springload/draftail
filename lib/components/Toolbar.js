import PropTypes from "prop-types";
import React from "react";

import ToolbarDefaults from "./ToolbarDefaults";
import ToolbarGroup from "./ToolbarGroup";

const Toolbar = (props) => {
  const { controls, getEditorState, onChange } = props;
  return (
    <div className="Draftail-Toolbar" role="toolbar">
      <ToolbarDefaults {...props} />

      <ToolbarGroup>
        {controls.map((Control, i) => (
          <Control
            // eslint-disable-next-line @thibaudcolas/cookbook/react/no-array-index-key
            key={i}
            getEditorState={getEditorState}
            onChange={onChange}
          />
        ))}
      </ToolbarGroup>
    </div>
  );
};

Toolbar.propTypes = {
  controls: PropTypes.array.isRequired,
  getEditorState: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
};

export default Toolbar;
