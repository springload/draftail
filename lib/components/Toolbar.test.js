import { OrderedSet } from "immutable";
import React from "react";
import { shallow } from "enzyme";

import Toolbar from "./Toolbar";

const mockProps = {
  currentStyles: new OrderedSet(),
  currentBlock: "unstyled",
  enableHorizontalRule: false,
  enableLineBreak: false,
  showUndoControl: false,
  showRedoControl: false,
  entityTypes: [],
  blockTypes: [],
  inlineStyles: [],
  controls: [],
  toggleBlockType: () => {},
  toggleInlineStyle: () => {},
  addHR: () => {},
  addBR: () => {},
  onUndoRedo: () => {},
  onRequestSource: () => {},
  getEditorState: () => {},
  onChange: () => {},
};

describe("Toolbar", () => {
  it("empty", () => {
    expect(shallow(<Toolbar {...mockProps} />)).toMatchSnapshot();
  });

  it("#controls", () => {
    expect(
      shallow(<Toolbar {...mockProps} controls={[() => <span>Test</span>]} />),
    ).toMatchSnapshot();
  });
});
