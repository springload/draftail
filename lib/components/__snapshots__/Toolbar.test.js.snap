// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar #blockTypes 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
>
  <Button
    active={false}
    icon={null}
    label="Test 1"
    name="TEST_1"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-2"
    label="Test 2"
    name="TEST_2"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-3"
    label="Test 3"
    name="TEST_3"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
</div>
`;

exports[`Toolbar #enableHorizontalRule 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
>
  <Button
    active={false}
    icon={null}
    label="HR"
    name="HORIZONTAL_RULE"
    onClick={[Function]}
    title={null}
  />
</div>
`;

exports[`Toolbar #enableLineBreak 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
>
  <Button
    active={false}
    icon={null}
    label="BR"
    name="BR"
    onClick={[Function]}
    title="shift + enter"
  />
</div>
`;

exports[`Toolbar #entityTypes 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
>
  <Button
    active={false}
    icon={null}
    label="Test 1"
    name="TEST_1"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-2"
    label="Test 2"
    name="TEST_2"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-3"
    label="Test 3"
    name="TEST_3"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
</div>
`;

exports[`Toolbar #inlineStyles 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
>
  <Button
    active={false}
    icon={null}
    label="Test 1"
    name="TEST_1"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-2"
    label="Test 2"
    name="TEST_2"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
  <Button
    active={false}
    icon="icon-test-3"
    label="Test 3"
    name="TEST_3"
    onClick={[Function]}
    title="No keyboard shortcut"
  />
</div>
`;

exports[`Toolbar empty 1`] = `
<div
  className="editor__toolbar"
  role="toolbar"
/>
`;
